Provide an option and a double-click gesture to show all local tracks for a process

julienw opened the pull request:
The goal for this patch is to provide various ways to easily display all local tracks for a process.

I tested with this profile: https://profiler.firefox.com/public/qsseyk7p9rz207y5eqnytyg1aap2hvdse6eky3g/

The initial request from @padenot was that when he hides then shows a track, all its local tracks should be displayed. But this has some drawbacks:
* if the process is the last one shown, clicking it won't hide it => therefore the manipulation won't work
* when testing it, I also realized that showing all tracks in a process can be cumbersome (a lot of new tracks will appear) and this can also be a performance problem with our current code.

Therefore I thought that we could do the same with a doubleclick instead. I added a small timeout for the simple click so that when double clicking we wouldn't first hide the process. As you'll see I had to change a few existing tests because of that, that's why I did this change in a separate commit at the end of the patch queue.

Because a doubleclick isn't discoverable easily, I also added options to the right click menus for the tracks.

You should look at the commits separately for an easier review. But do not look at commits when their log starts with `[Do not review`.

Here are the changes:
* added an item to both global track's and local track's context menu
* double clicking a global track in any context menu or the track popup will display all of its local tracks

[production](https://profiler.firefox.com/public/qsseyk7p9rz207y5eqnytyg1aap2hvdse6eky3g/)
[deploy preview](https://deploy-preview-4286--perf-html.netlify.app/public/qsseyk7p9rz207y5eqnytyg1aap2hvdse6eky3g/)

padenot commented:
This works well for my needs, I just tested the deployed preview, thanks a lot!

julienw commented on src/components/timeline/TrackContextMenu.js:
```diff
@@ -993,12 +1034,15 @@ class TimelineTrackContextMenuImpl extends PureComponent<
-      isolateScreenshot ? (
+      isolateScreenshot ||
+      showLocalTracksInProcess ||
+      hideTrack ? (
```
I've seen that `hideTrack` wasn't present here: you'll see this change in a test later too, but I couldn't understand in which case only `hideTrack` would be displayed -- it's possible because this happens in a test, but couldn't reproduce with a real profile.

julienw commented on src/test/components/TrackContextMenu.test.js:
```diff
@@ -721,8 +722,8 @@ describe('timeline/TrackContextMenu', function () {
   describe('when a local track is right clicked', function () {
-    function setupLocalTrack() {
-      const results = setup();
+    function setupLocalTrack(profile) {
+      const results = setup(profile);
```
After all I don't use this change, but still kept it because this does no harm...

julienw commented on src/test/components/TrackContextMenu.test.js:
```diff
@@ -825,6 +826,84 @@ describe('timeline/TrackContextMenu', function () {
+
+    // This runs 2 tests: the first right clicks a global track, the second
+    // right clicks the local track.
+    it.each([
+      { type: 'global', trackIndex: 0 },
+      {
+        type: 'local',
+        pid: 1000,
+        trackIndex: 0,
+      },
```
Initially I wrote 2 tests, but they were nearly identical, and I thought it was more readable to group them this way, even though we don't do that usually... Tell me what you think.

julienw replied:
yeah, I thought that this way it's more obvious what difference the tests have. Otherwise it was a bit buried in the code and I couldn't find another way that I found better...

canova commented on src/actions/profile-view.js:
```diff
@@ -1171,6 +1171,28 @@ export function showGlobalTrack(trackIndex: TrackIndex): ThunkAction<void> {
+ * This action shows a specific global track and its local tracks.
+ */
+export function showGlobalTrackIncludingLocalTracks(
+  trackIndex: TrackIndex,
+  pid: Pid
```
Curious we need both the trackIndex and pid. Isn't one of them enough?

Edit: I see it now that we use it for directly making the local tracks visible with `hiddenLocalTracksByPid.set(action.pid, new Set());` below. Maybe we can still get this information from the track reference, but probably not worth the effort that much.

julienw replied:
The main problem is that in reducers we don't have access to all the state. We need to provide in actions all data needed to handle actions in reducers. That's different to what happens in thunk actions where we have access to all the state. But I believe it's usually a better model, because the state is more self-contained this way, than in thunk actions where everything can interact with everything.

canova commented on src/test/components/TrackContextMenu.test.js:
```diff
@@ -825,6 +826,84 @@ describe('timeline/TrackContextMenu', function () {
+        pid: 1000,
+        trackIndex: 0,
+      },
+    ])(`from the $type track's context menu`, (rightClickedTrackReference) => {
```
oh I didn't know that you could write like `$type`, nice!

julienw replied:
Note that this isn't JavaScript syntax, this comes with the `test.each` jest API.

canova replied:
oh yeah, I assumed so :)

canova commented on src/components/timeline/TrackContextMenu.js:
```diff
@@ -248,18 +249,27 @@ class TimelineTrackContextMenuImpl extends PureComponent<
+      } else {
+        hideGlobalTrack(trackIndex);
+      }
+    }, 80);
```
After testing, it looks like 80 is pretty good for me

canova reviewed (approved):
Looks pretty good to me, thanks Julien!
Also did some manual testing with 80ms delay and it doesn't feel disturbing at all. Also from the perfnow, they recommended below 100ms delay for the best, and 200ms for optimum/okay one.

julienw commented:
I fixed the review comment, moving the new item below and even adding a separator when needed. I also added some more logic so that the item isn't added when there's no local tracks, and it's disabled if all local tracks are visible already. I won't ask for a new review for these changes because they're fairly small and Nazim is quite overloaded already.
